### PR TITLE
Clean state if it doesn't belong to country

### PR DESCRIPTION
**Pull request by vladstoick**
Right now if validation isn't required, there's no clean-up done. This can lead to state being left over when changing countries.

On our solidus store we have `Spree::Config[:address_requires_state]` disabled and we only have `country.states_required = true` for countries for which we have data so we can show a select element. The current address validation for state isn't executed if `country.states_required` is false or `Spree::Config[:address_requires_state]` is false. 

This leads to a problem: when changing from a country that has states to another without states, the `state_id` remains on the address. 

This PR changes the state_validate to clean-up states.

### CODE REVIEW

File: core/app/models/spree/address.rb [hunk 1]
```diff
@@ -157,11 +157,17 @@ def country_iso=(iso)
...
-        # Skip state validation without country (also required)
-        # or when disabled by preference
-        return if country.blank? || !Spree::Config[:address_requires_state]
-        return unless country.states_required
+        # Skip state validation without country
+        return if country.blank?
 
+        # If states are not required we still check if state belongs to country
+        # If it doesn't we just remove it without adding state errors
+        unless country.states_required || readonly?
```
Comment (magnusvk) [inline core/app/models/spree/address.rb]: Are you familiar with the immutable address work? I'm not sure this will work as you expect it to be, [as all persisted addresses are immutable](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/address.rb#L147). It feels to me that there's a better place for this logic, perhaps [here](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/address.rb#L46). cc @gvaughn also who may want to take a look at this.

Comment (vladstoick) [inline core/app/models/spree/address.rb]: I've added the readonly check thinking it was just some special case, not knowing about the immutability part since our store is still on 1.0. However, if we remove it all specs pass except this one. 
```
  1) Spree::CheckoutController#update fails to transition to complete from confirm when the country is not a shippable country due to no available shipping rates for any of the shipments
     Failure/Error: order.update_attributes(state: 'confirm')
     ActiveRecord::ReadOnlyRecord:
       Spree::Address is marked as readonly
     # ./spec/controllers/spree/checkout_controller_spec.rb:358:in `block (5 levels) in <top (required)>'
``` which doesn't seem related at all.

Putting in the merge function doesn't seem explicit it at all, and the validation still runs when creating the new address.

Comment (gvaughn) [inline core/app/models/spree/address.rb]: @magnusvk Agreed. #immutable_merge should look at `country.states_required` to potentially leave the state_id out of the new_address
